### PR TITLE
builtin: Add `__iter__` method to `ArrayAttr`

### PR DESCRIPTION
**Pull request by AntonLydike**
This adds the `__iter__` method to the `ArrayAttr`, making it itself be an `Iterable[Attribute]`. This will significantly reduce boilerplate code to interact with ArrayAttrs.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
not: test?

**Comment by AntonLydike** [comment]
> nit: test?

Added a small test!

File: xdsl/irdl.py [hunk 1]
```diff
@@ -219,7 +219,9 @@ def irdl_to_attr_constraint(
 
     # Attribute class case
     # This is a coercion for an `BaseAttr`.
-    if isclass(irdl) and issubclass(irdl, Attribute):
+    if isclass(irdl) \
+            and not isinstance(irdl, GenericAlias) \
```
Comment (PapyChacal) [inline xdsl/irdl.py:223]: What's the impact of this additional condition, and its link to `ArrayAttr` being `Iterable`?

Comment (AntonLydike) [reply xdsl/irdl.py:223]: Not exactly sure, it seems that having generic + multiple inheritance (possibly from `Iterable` also not being a class in the classical sense) causes the type to become a `GenericAlias`, which satisfies `isclass`, but cannot be passed to `issubclass`... This fixes it for now, but @math-fehr might have a better solution down the road.

**Review by webmiche — COMMENTED** [review]
I think we had a discussion like this before where we decided to not expose the data without having to access the `.data` field. AFAIR, the reason was that we prefered the explicitness/clarity of what you access. Not sure how that ties into this discussion. I think @superlopuh had some opinions on that.

File: xdsl/irdl.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/irdl.py:223]: I don't know this corner of IRDL well, so I'm not sure, but it reads to me like it's plainly ignoring some constraints here. Maybe I'm just missing something?

**Comment by webmiche** [comment]
In that case, I am fine with this change. If it ends up being confusing, we can still remove it easily...

**Review by webmiche — APPROVED** [review]

**Review by georgebisbas — APPROVED** [review]

**Comment by math-fehr** [comment]
Before merging, I'll just fix the `isa` function.
I'll have more time tomorrow/friday though!

**Pull request merged**